crates/fluvio-protocol/src/core/bytebuf.rs

```diff
@@ -0,0 +1,236 @@
+
+        if len < 1 {
+            return Ok(());
+        }
+
+        self.inner = src.copy_to_bytes(len as usize);
+
+        Ok(())
+    }
+}
+
+impl Encoder for ByteBuf {
+    fn write_size(&self, version: Version) -> usize {
+        0_u32.write_size(version) + self.inner.len()
+    }
+
+    fn encode<T>(&self, dest: &mut T, version: Version) -> Result<(), Error>
+    where
+        T: BufMut,
+    {
+        let remaining = dest.remaining_mut();
+        let expected = self.write_size(version);
+
+        if remaining < expected {
+            return Err(Error::new(
+                ErrorKind::UnexpectedEof,
+                format!(
+                    "Not enough capacity for ByteBuf. Expected: {}, Remaining: {}",
+                    expected, remaining
+                ),
+            ));
+        }
+
+        dest.put_u32(self.inner.len() as u32);
+        dest.put(self.inner.clone());
+
+        Ok(())
+    }
+
+    fn as_bytes(&self, _version: Version) -> Result<Bytes, Error> {
+        Ok(self.inner.clone())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::io::Cursor;
+
+    use crate::{Decoder, Encoder};
+    use super::ByteBuf;
+
+    #[test]
+    fn test_is_empty_by_default() {
+        let bytebuf = ByteBuf::default();
+
+        assert!(bytebuf.is_empty());
+        assert!(bytebuf.inner.is_empty());
+    }
+
+    #[test]
+    fn test_is_empty_from_inner_vec() {
+        let bytebuf = ByteBuf::from(vec![128, 129, 130, 131]);
+
+        assert_eq!(bytebuf.is_empty(), bytebuf.inner.is_empty());
+    }
+
+    #[test]
+    fn test_is_empty_from_inner_empty_vec() {
+        let bytebuf = ByteBuf::from(Vec::default());
+
+        assert_eq!(bytebuf.is_empty(), bytebuf.inner.is_empty());
+    }
+
+    #[test]
+    fn test_len_from_inner_vec() {
+        let bytebuf = ByteBuf::from(vec![128, 129, 130, 131]);
+
+        assert_eq!(bytebuf.len(), bytebuf.inner.len());
+    }
+
+    #[test]
+    fn test_encode_bytebuf() {
+        let mut dest = Vec::default();
+        let value: ByteBuf = ByteBuf::from(vec![0x10, 0x11]);
+        let result = value.encode(&mut dest, 0);
+
+        assert!(result.is_ok());
+
+        // Length + Contents
+        assert_eq!(dest.len(), 6);
+
+        // Length Bytes
+        assert_eq!(dest[0], 0x00);
+        assert_eq!(dest[1], 0x00);
+        assert_eq!(dest[2], 0x00);
+        assert_eq!(dest[3], 0x02);
+
+        // Actual Content
+        assert_eq!(dest[3], 0x02);
+        assert_eq!(dest[5], 0x11);
+
+        // Length in 32bits (4 bytes) + Contents (5 elements)
+        assert_eq!(value.write_size(0), dest.len());
+    }
+
+    #[test]
+    fn test_bytebuf_encodes_transparent_with_vecu8() {
+        let raw_data: [u8; 10] = [0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A];
+
+        // Data encoded for Vec<u8> will have the first 4 bytes (32 bits) for
+        // data length, and the remaining bytes represents the actual data
+        let encoded_expect: [u8; 14] = [
```

Comment:
maybe instead of explicitly defining encoded_expect, we could use: `Vec::<u8>::from(raw_data).encode(&mut encoded_data, 0).expect("vec encoding failed");`